pkg/daemon/update.go

```diff
@@ -880,14 +885,27 @@ func (dn *Daemon) updateOnClusterBuild(oldConfig, newConfig *mcfgv1.MachineConfi
 	// If the new image pullspec is already on disk, do not attempt to re-apply
 	// it. rpm-ostree will throw an error as a result.
 	// See: https://issues.redhat.com/browse/OCPBUGS-18414.
-	if oldImage != newImage && newImage != "" {
+	if oldImage != newImage {
+		// If the new image field is empty, set it to the OS image URL value
+		// provided by the MachineConfig to do a rollback.
+		if newImage == "" {
+			klog.Infof("%s empty, reverting to osImageURL %s from MachineConfig %s", constants.DesiredImageAnnotationKey, newConfig.Spec.OSImageURL, newConfig.Name)
+			newImage = newConfig.Spec.OSImageURL
+		}
 		if err := dn.updateLayeredOSToPullspec(newImage); err != nil {
 			return err
 		}
 	} else {
 		klog.Infof("Image pullspecs equal, skipping rpm-ostree rebase")
 	}
 
+	// If the new OS image equals the OS image URL value, this means we're in a
+	// revert-from-layering situation. This also means we can return early after
+	// taking a different path.
+	if newImage == newConfig.Spec.OSImageURL {
```

Comment:
Question: so the conditions to indicate a revert are either:
1. an empty desired image field, or
2. setting the desired image back to the current OSImageURL of the cluster payload

Is that correct? If so, could you help me understand when the second case would happen? (i.e. how desiredImage == renderedConfig.spec.osimageurl)

Reply:
So, the revert operation should only happen if the `desiredImage` annotation is set to empty and `currentImage` is still set to something. In that case, we specifically set `newImage` to equal `newConfig.Spec.OSImageURL` a few lines above this block (on line 893).